hibernate-core/src/main/java/org/hibernate/internal/util/ReflectHelper.java

```diff
@@ -229,9 +234,17 @@ private static Getter getter(Class clazz, String name) throws MappingException {
 		return PropertyAccessStrategyMixedImpl.INSTANCE.buildPropertyAccess( clazz, name ).getGetter();
 	}
 
-	public static Object getConstantValue(String name, ClassLoaderService classLoaderService) {
+	public static Object getConstantValue(String name, SessionFactoryImplementor factory) {
+		boolean conventionalJavaConstants = factory.getSessionFactoryOptions().isConventionalJavaConstants();
 		Class clazz;
 		try {
+			if ( conventionalJavaConstants &&
+				!Pattern.compile( JAVA_CONSTANT_PATTERN )
```

Comment:
Actually just verified that that absolutely should be the case.  From Pattern Javadoc:

````
/**
 *...
 * Instances of this class are immutable and are safe for use by multiple concurrent threads. Instances of the Matcher class are not safe for such use.
 *...
 */